internal/pkg/graph/graph.go

```diff
@@ -215,3 +313,122 @@ func TopologicalOrder[V comparable](digraph *Graph[V]) (*TopologicalSorter[V], e
+		adjacentVertexSkipStatus: adjacentVertexSkipStatus,
+		processVertex:            processVertexFunc,
+	}
+	return traversal.execute(ctx, g)
+}
+
+/*
+DownwardTraversal performs a downward traversal on the graph starting from root nodes (nodes with no parents)
+and moving towards leaf nodes (nodes with parents). It applies the specified process function to each
+vertex in the graph, skipping vertices with the "adjacentVertexSkipStatus" status, and continuing traversal
+until reaching vertices with the "requiredVertexStatus" status.
+The traversal is concurrent and may process vertices in parallel.
+Returns an error if the traversal encounters any issues.
+*/
+func (g *Graph[V]) DownwardTraversal(ctx context.Context, processVertexFunc func(context.Context, V) error, adjacentVertexSkipStatus, requiredVertexStatus string) error {
+	traversal := &graphTraversal[V]{
+		mu:                       sync.Mutex{},
+		seen:                     make(map[V]struct{}),
+		findBoundaryVertices:     func(g *Graph[V]) []V { return g.Roots() },
+		findAdjacentVertices:     func(g *Graph[V], v V) []V { return g.children(v) },
+		filterVerticesByStatus:   func(g *Graph[V], v V, status string) []V { return g.filterParents(v, status) },
+		requiredVertexStatus:     requiredVertexStatus,
+		adjacentVertexSkipStatus: adjacentVertexSkipStatus,
+		processVertex:            processVertexFunc,
+	}
+	return traversal.execute(ctx, g)
+}
+
+type graphTraversal[V comparable] struct {
+	mu                       sync.Mutex
+	seen                     map[V]struct{}
+	findBoundaryVertices     func(*Graph[V]) []V
+	findAdjacentVertices     func(*Graph[V], V) []V
+	filterVerticesByStatus   func(*Graph[V], V, string) []V
+	requiredVertexStatus     string
+	adjacentVertexSkipStatus string
+	processVertex            func(context.Context, V) error
+}
+
+func (t *graphTraversal[V]) execute(ctx context.Context, graph *Graph[V]) error {
+
+	ctx, cancel := context.WithCancel(ctx)
+	defer cancel()
+
+	vertexCount := len(graph.vertices)
+	if vertexCount == 0 {
+		return nil
+	}
+	eg, ctx := errgroup.WithContext(ctx)
+	vertexCh := make(chan V, vertexCount)
+	defer close(vertexCh)
+
+	processVertices := func(ctx context.Context, graph *Graph[V], eg *errgroup.Group, vertices []V, vertexCh chan V) {
+		for _, vertex := range vertices {
+			vertex := vertex
+			// Delay processing this vertex if any of its dependent vertices are yet to be processed.
+			if len(t.filterVerticesByStatus(graph, vertex, t.adjacentVertexSkipStatus)) != 0 {
+				continue
+			}
+			if !t.markAsSeen(vertex) {
+				// Skip this vertex if it's already been processed by another routine.
+				continue
+			}
+			eg.Go(func() error {
+				if err := t.processVertex(ctx, vertex); err != nil {
+					return err
+				}
+				// Assign new status to the vertex upon successful processing.
+				graph.updateStatus(vertex, t.requiredVertexStatus)
+				vertexCh <- vertex
+				return nil
+			})
+		}
+	}
+
+	eg.Go(func() error {
```

Comment:
This looks great and more efficient than processing one by one. Thank you for diving deep into this implementation. My only concern is if it really makes that much difference from UX perspective to compromise the extra impl complexity we introduce, which potentially lower the maintainability and make it more error-prone 💭

Reply:
I need the concurrent functionality because 
https://github.com/aws/copilot-cli/pull/5509/files#diff-e21274f5c1fd77d8e5391a27014297ea2818cc1055df6c7a0eeff56e3251aebaR245-R260

1. Build dependency graph based on dependencies.
2. When calling `UpwardTraversal` function we are send this as `ProcessVertexFunc`
      a. waitForDependencies of a container.
      b. Run the container.
      c. Wait for container to Start.
      
Incase if there are no dependencies for `A` and `B`,  if we do not use go routines in the graph pkg, 
I can not achieve the desired functionality of container A and B to start in parallel.

Reply:
Gotcha that makes sense to me.